src/app/loadout-analyzer/hooks.tsx

```diff
@@ -0,0 +1,138 @@
+import { savedLoadoutParametersSelector } from 'app/dim-api/selectors';
+import {
+  allItemsSelector,
+  createItemContextSelector,
+  unlockedPlugSetItemsSelector,
+} from 'app/inventory/selectors';
+import { DimStore } from 'app/inventory/store-types';
+import { getAutoMods } from 'app/loadout-builder/process/mappers';
+import { Loadout } from 'app/loadout-drawer/loadout-types';
+import { d2ManifestSelector } from 'app/manifest/selectors';
+import { currySelector } from 'app/utils/selectors';
+import { noop } from 'lodash';
+import {
+  ReactNode,
+  createContext,
+  useCallback,
+  useContext,
+  useEffect,
+  useId,
+  useState,
+  useSyncExternalStore,
+} from 'react';
+import { useSelector } from 'react-redux';
+import { createSelector } from 'reselect';
+import { LoadoutBackgroundAnalyzer } from './store';
+import { LoadoutAnalysisContext, LoadoutAnalysisResult, LoadoutAnalysisSummary } from './types';
+
+const LoadoutAnalyzerReactContext = createContext<LoadoutBackgroundAnalyzer | null>(null);
+
+const autoModSelector = createSelector(
+  d2ManifestSelector,
+  unlockedPlugSetItemsSelector.selector,
+  (defs, unlockedPlugSetItems) => defs && getAutoMods(defs, unlockedPlugSetItems)
+);
+
+// It'd be really neat if this one didn't depend on the storeId but
+// unlockedPlugSetItemsSelector and as a result, autoModSelector
+// do need it.
+const autoOptimizationContextSelector = currySelector(
+  createSelector(
+    createItemContextSelector,
+    unlockedPlugSetItemsSelector.selector,
+    allItemsSelector,
+    savedLoadoutParametersSelector,
+    autoModSelector,
+    (itemCreationContext, unlockedPlugs, allItems, savedLoLoadoutParameters, autoModDefs) =>
+      itemCreationContext.defs &&
+      autoModDefs &&
+      ({
+        itemCreationContext,
+        unlockedPlugs,
+        savedLoLoadoutParameters,
+        allItems,
+        autoModDefs,
+      } satisfies LoadoutAnalysisContext)
+  )
+);
+
+/** Wrapper component that holds the analyzer task and results. */
+export function MakeLoadoutAnalysisAvailable({ children }: { children: ReactNode }) {
+  const [analyzer, setAnalyzer] = useState<LoadoutBackgroundAnalyzer | null>(null);
+  useEffect(() => {
+    setAnalyzer(new LoadoutBackgroundAnalyzer());
+    return () => {
+      setAnalyzer((oldAnalyzer) => {
+        oldAnalyzer?.destroy();
+        return null;
+      });
+    };
+  }, []);
```

Comment:
That's what I did at first, but the React people are very adamant about the fact that [`useEffect` with an empty dependency array guarantees nothing and your setup and cleanup logic need to be symmetrical](https://react.dev/reference/react/useEffect#my-cleanup-logic-runs-even-though-my-component-didnt-unmount).

Reply:
Like, we did turn off `StrictMode` because it risked adding bugs to prod, but the "effect cleanup logic with no setup logic" approach will immediately break in Strict Mode. I don't really like it either but I don't know of a better way to manage an external resource without a symmetric effect/cleanup.

Reply:
I guess it's on them to build something that allows for this pattern without forcing a double-render then...